src/Make/MakeICal.php

```diff
@@ -0,0 +1,289 @@
+use Elabftw\Models\Config;
+use Elabftw\Models\Items;
+use Elabftw\Models\Scheduler;
+use Elabftw\Models\Users;
+use Elabftw\Models\Todolist;
+use Elabftw\Models\UnfinishedSteps;
+use Elabftw\Models\Notifications\StepDeadline;
+use Elabftw\Models\Steps;
+use Eluceo\iCal\Domain\Entity\Calendar;
+use Eluceo\iCal\Domain\Entity\Event;
+use Eluceo\iCal\Domain\Entity\Todo;
+use Eluceo\iCal\Domain\ValueObject\Alarm;
+use Eluceo\iCal\Domain\ValueObject\Alarm\DisplayAction;
+use Eluceo\iCal\Domain\ValueObject\Alarm\RelativeTrigger;
+use Eluceo\iCal\Domain\ValueObject\DateTime;
+use Eluceo\iCal\Domain\ValueObject\TimeSpan;
+use Eluceo\iCal\Domain\ValueObject\UniqueIdentifier;
+use Eluceo\iCal\Presentation\Factory\CalendarFactory;
+
+use function array_merge;
+use function date_default_timezone_get;
+use function json_decode;
+use function sprintf;
+use function strlen;
+
+/**
+ * Make an iCal file
+ */
+class MakeICal extends AbstractMake implements StringMakerInterface
+{
+    protected string $contentType = 'text/calendar; charset=utf-8';
+
+    public function __construct(protected string $calendarToken)
+    {
+        parent::__construct();
+    }
+
+    /**
+     * Create an iCal file from events
+     */
+    public function getFileContent(): string
+    {
+        $iCalString = $this->getICalString($this->getEntries($this->getCalendarProperties()));
+        $this->contentSize = strlen($iCalString);
+        return $iCalString;
+    }
+
+    /**
+     * Return a nice name for the file
+     */
+    public function getFileName(): string
+    {
+        return 'eLabFTW-calendar.ics';
+    }
+
+    private function getCalendarProperties(): array
+    {
+        $sql = 'SELECT `team`, `created_by`,
+                        `all_events`,
+                        JSON_ARRAYAGG(`category`) AS `categories`,
+                        JSON_ARRAYAGG(`item`) AS `items`,
+                        `todo`, `unfinished_steps_scope`
+                    FROM `calendars`
+                    LEFT JOIN `calendar2items_types` ON `calendars`.`id` = `calendar2items_types`.`calendar`
+                    LEFT JOIN `calendar2items` ON `calendars`.`id` = `calendar2items`.`calendar`
+                    WHERE `token` LIKE :token
+                    GROUP BY `calendars`.`id`
+                    LIMIT 1';
+        $req = $this->Db->prepare($sql);
+        $req->bindParam(':token', $this->calendarToken);
+        $this->Db->execute($req);
+        $res = $this->Db->fetch($req);
+        foreach (array('categories', 'items') as $key) {
+            $res[$key] = json_decode($res[$key], true);
+            if ($res[$key][0] === null) {
+                $res[$key] = null;
+            }
+        }
+        return $res;
+    }
+
+    private function getEntries(array $calendarProps): array
+    {
+        $entries = array();
+        // read an individual event
+        // can be used to export a single event
+        // if (isset($calendarProps['event'])) {
```

Comment:
is this commented code supposed to be left here?

Reply:
It could be used to download individual events via a download button that has yet to be implemented.

It can also be deleted if such a button is considered not useful.

Reply:
What about adding `format=ical` to a `GET /api/v2/event/{id}`?